Physics Controller Implementation Part 1 - POSITION

Weetsy opened the pull request:
# Changes
This is part 1 of the PhysicsController implementation. This first part only includes the PIPELINE stage for physics calculations.

Changelog:
* Added the PhysicsController as a studious controller. 
* Added the first step in the physics pipeline (POSITION).
* Added unit tests to validate physics controller functionality.
* Wired physics controller into GameInstance.
* Added physics sample code to 3D demo. Press backspace to fire projectiles.
* Changed Polygon to shared_ptr in ModelImport. The current implementation had scary memory lifetime problems when instantiating 3D objects from separate threads.
* Decoupled some 3D demo headers from the actual engine. 
### Context
<!--- Give a brief description of your changes. -->
One of the missing key features to studious engine right now is the lack of a global physics system. These changes introduce the starting point for the global physics engine. There will be more upcoming PRs that introduce the later pipeline stages. The pipeline, once completed, will have the following stages:

POSITION -> COLLISION -> FINALIZE -> REPORT. 

See my inline comment above the updatePosition method for more information. 

### Issues
<!--- List any relavant GitHub issues this PR is addressing here. -->
#20 
### Considerations
<!--- If any major decisions, breaking changes, or redesigns were made, describe them here. Give a brief summary describing how you came to this conclusion. -->
There are some breaking changes that will require some light game code refactoring:

* PhysicsController needs to be defined at a global level like AnimationController and passed into the GameInstance constructor. See the 3D demo code for examples on how to do this. 
* game.hpp and inputMonitor.hpp are no longer part of the game engine. If you want these in your game projects, you will need to copy them over from the 3D example folder. 
## QA Checklist

- [x] I ran `cpplint --linelength=120 --exclude=src/main/opengl --recursive src/main/` and corrected any issues.
- [x] I added unit tests to relevant code.
- [ ] ~I added/revised Doxygen documentation on new code.~
- [x] I can compile using G++.
- [x] I am passing all unit tests. (`./setupBuild.sh -t`)
- [x] I updated the basic template project if necessary (https://github.com/Weetsy/studious-template) [PhysicsController update found here.](https://github.com/Weetsy/studious-template/pull/1)

alec-jackson commented:
Are you thinking the controls will just be setting p,v,a or are you going to support force vectors

alec-jackson commented on src/main/engine/Misc/src/physics.cpp:
```diff
@@ -4,12 +4,72 @@
+         *          2
+         */
+        // Use OpenCL to speed this up maybe? - move this to PhysicsObject as a method
+        auto distFunc = [] (vec3 position, vec3 velocity, vec3 acceleration, float deltaTime) {
```
I'm trying to think of cases where direct manipulation of acceleration is important, but I can't think of anything that shouldn't be done with forces. I think gravity would be the most obvious case, but we should probably encourage gravity to be done with force.

alec-jackson commented:
> > Are you thinking the controls will just be setting p,v,a or are you going to support force vectors
> 
> I'm thinking of allowing the user to set p/v but not a. I'm thinking of having some kind of Fnet value we could use to get acceleration from. I'm not sure if we should keep a vector of different forces acting on an object, or just keep a single Fnet vector in the PhysicsObject that we can add/subtract directly. I'm going to mess around with different options here to see what works better, and I'll update this PR once I have something I think is reasonable.

The tricky part seems like it would be the short term forces. The longer term ones can be applied and changed over time intervals and the actual tick time isn't very important, but things like collisions (short term forces) seem like they need to be handled as shorter than a regular time interval would be guaranteed to be. I'm seeing 2 options off the top of my head. When a force is applied you can make it an "instantaneous force" or a "continuous force". Which an instantaneous force would get cleared after the first tick. We'd probably want to impose a min and max tick time to bound it, but it would leave some variability. Alternatively for instant forces (e.g. collisions) you could force those to programmatically set a collision time that it would expire after. Then when updating it a partial force could be applied for the remainder of the force time between physics ticks.

Sort of just some random thoughts. I need to fully look through what you've implemented here to know if you already have a solution for these problems.

Weetsy commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -13,7 +13,6 @@
 
 - [ ] I ran `cpplint --linelength=120 --exclude=src/main/opengl --recursive src/main/` and corrected any issues.
 - [ ] I added unit tests to relevant code.
-- [ ] I added/revised Doxygen documentation on new code.
```
Removing this as a QA requirement until we declare studious engine v1.0.0. It's too much of a hassle to keep this stuff up to date when it keeps inevitably changing...

alec-jackson replied:
Agreed, in the future I think it'd be a better practice to do changes like this as their own small PRs

Weetsy commented on setupBuild.sh:
```diff
@@ -8,6 +8,8 @@ runTests=false
 buildAll=false
 installLib=false
+filter_tests=false
+physThreads=1
```
With these changes you can now specify which GTests to use, and how many threads you want to allocate to the physics engine.

Weetsy commented:
@alec-jackson This is ready for review again. I scrapped the idea I described above as it over-complicated the physics calculation and went with the instantaneous force apply method.

alec-jackson commented on CMakeLists.txt:
```diff
@@ -11,6 +11,9 @@ endif()
   add_compile_options(-DGFX_EMBEDDED)
 endif()
+if (PHYS_THREADS)
+  add_compile_options(-DPHYS_THREADS=${PHYS_THREADS})
```
I feel like we shouldn't do this as a compile option. 

What about something like

if (physThreadConf){
}else if (physThreadCmd){
} else {
}

alec-jackson replied:
Got distracted looking at amps and didn't finish the review. 

It'd take some research to know the equivilent nproc on all the platforms we're targeting. Maybe that's something we should open as an issue off this PR.

alec-jackson commented on src/main/engine/AnimationController/test/src/AnimationControllerTests.cpp:
```diff
@@ -747,9 +748,9 @@ TEST_F(GivenAnAnimationControllerReady, WhenKeyFramesAddedForMultipleObjects_The
     float targetTime = 3.0f;
     deltaTime = 1.0f;
     vec3 expectedTransformation_1 = originalPosition_1 +
-        ((desiredPosition - originalPosition_1) * vec3((deltaTime * 1) / targetTime));
```
...LMFAO

Weetsy commented:
@alec-jackson Okay, so one thing lead to another, and another... I changed how configurations are loaded into studious. The existing code was pretty scary, so I updated it and made it handle any passed in config option. That config is passed into GameInstance to avoid excessive game breaking changes whenever a config is added. I also moved controller creation over to the GameInstance class, so the user doesn't have to worry about that.

I implemented the default physics thread stuff we talked about. It looks like in C++ you can get the number of logical processors on your system with `std::thread::hardware_concurrency()`, which may fail sometimes. If it fails, then we fallback to PHYS_THREADS.
